src/mesh_tools.rs

```diff
@@ -0,0 +1,200 @@
+use std::collections::HashMap;
+
+use smallvec::SmallVec;
+
+use crate::convert::{cast_u32, cast_usize};
+use crate::geometry::Geometry;
+use crate::mesh_topology_analysis::face_to_face_topology;
+
+/// Crawls the geometry to find continuous patches of geometry.
+/// Returns a vector of new separated geometries.
+#[allow(dead_code)]
+pub fn separate_isolated_meshes(geometry: &Geometry) -> Vec<Geometry> {
+    let face_to_face = face_to_face_topology(geometry);
+    let mut available_face_indices: Vec<u32> =
+        face_to_face.keys().map(|key| cast_u32(*key)).collect();
+    let mut patches: Vec<Geometry> = Vec::new();
+
+    while let Some(first_face_index) = available_face_indices.pop() {
+        let mut separate_face_indices: Vec<u32> = Vec::new();
+
+        crawl_faces(
+            first_face_index,
+            &face_to_face,
+            &mut available_face_indices,
+            &mut separate_face_indices,
+        );
+
+        patches.push(
+            Geometry::from_faces_with_vertices_and_normals_remove_orphans(
+                separate_face_indices
+                    .iter()
+                    .map(|face_index| geometry.faces()[cast_usize(*face_index)]),
+                geometry.vertices().to_vec(),
+                geometry.normals().to_vec(),
+            ),
+        );
+    }
+
+    patches
+}
+
+fn crawl_faces(
+    current_face_index: u32,
+    face_to_face: &HashMap<u32, SmallVec<[u32; 8]>>,
+    available_face_indices: &mut Vec<u32>,
+    separate_face_indices: &mut Vec<u32>,
+) {
+    separate_face_indices.push(current_face_index);
+
+    if let Some(all_neighbors_indices) = face_to_face.get(&current_face_index) {
+        let neighbor_indices: Vec<_> = all_neighbors_indices
+            .iter()
+            .copied()
+            .filter(|n| available_face_indices.iter().any(|a_f| a_f == n))
+            .collect();
+        available_face_indices.retain(|f| neighbor_indices.iter().all(|n| n != f));
+        for neighbor_index in neighbor_indices {
+            crawl_faces(
+                neighbor_index,
+                face_to_face,
+                available_face_indices,
+                separate_face_indices,
+            );
+        }
+    }
+}
```

Comment:
I have a played a bit with optimization here and the outline looks like this: https://gist.github.com/ondrowan/d635981610ed0ff9394f56c3a81f6809

It uses depth-first search which removes the need for recursion and vector allocations. It's however not finished and there is space for improvement, as seen in notes.